lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1434 @@
+  if (auto dstBundle = dstType.dyn_cast<BundleType>()) {
+    auto srcBundle = srcType.cast<BundleType>();
+    for (unsigned dstIdx = 0, e = dstBundle.getNumElements(); dstIdx < e;
+         ++dstIdx) {
+      auto dstField = dstBundle.getElements()[dstIdx].name.getValue();
+      auto srcIdx = srcBundle.getElementIndex(dstField);
+      if (!srcIdx)
+        continue;
+      auto &dstElt = dstBundle.getElements()[dstIdx];
+      auto &srcElt = srcBundle.getElements()[*srcIdx];
+      if (dstElt.isFlip) {
+        traceResets(srcElt.type, src, srcID + srcBundle.getFieldID(*srcIdx),
+                    dstElt.type, dst, dstID + dstBundle.getFieldID(dstIdx),
+                    loc);
+      } else {
+        traceResets(dstElt.type, dst, dstID + dstBundle.getFieldID(dstIdx),
+                    srcElt.type, src, srcID + srcBundle.getFieldID(*srcIdx),
+                    loc);
+      }
+    }
+  } else if (auto dstVector = dstType.dyn_cast<FVectorType>()) {
+    auto srcVector = srcType.cast<FVectorType>();
+    auto srcElType = srcVector.getElementType();
+    auto dstElType = dstVector.getElementType();
+    // Collapse all elements into one shared element. See comment in traceResets
+    // above for some context.
+    traceResets(dstElType, dst, dstID + dstVector.getFieldID(0), srcElType, src,
+                srcID + srcVector.getFieldID(0), loc);
+  } else if (dstType.isGround()) {
+    if (dstType.isa<ResetType>() || srcType.isa<ResetType>()) {
+      FieldRef dstField(dst, dstID);
+      FieldRef srcField(src, srcID);
+      LLVM_DEBUG(llvm::dbgs() << "Visiting driver '" << getFieldName(dstField)
+                              << "' = '" << getFieldName(srcField) << "' ("
+                              << dstType << " = " << srcType << ")\n");
+
+      // Determine the leaders for the dst and src reset networks before we make
+      // the connection. This will allow us to later detect if dst got merged
+      // into src, or src into dst.
+      ResetSignal dstLeader =
+          *resetClasses.findLeader(resetClasses.insert({dstField, dstType}));
+      ResetSignal srcLeader =
+          *resetClasses.findLeader(resetClasses.insert({srcField, srcType}));
+
+      // Unify the two reset networks.
+      ResetSignal unionLeader = *resetClasses.unionSets(dstLeader, srcLeader);
+      assert(unionLeader == dstLeader || unionLeader == srcLeader);
+
+      // If dst got merged into src, append dst's drives to src's, or vice
+      // versa.
+      auto &unionDrives = resetDrives[unionLeader];
+      if (dstLeader != srcLeader) {
+        if (unionLeader == dstLeader)
+          unionDrives.append(std::move(resetDrives[srcLeader]));
+        else
+          unionDrives.append(std::move(resetDrives[dstLeader]));
+      }
+
+      // Keep note of this drive so we can point the user at the right location
+      // in case something goes wrong.
+      unionDrives.push_back({{dstField, dstType}, {srcField, srcType}, loc});
+    }
```

Comment:
The use of equivalence classes here is _extremely_ elegant.

Reply:
Yeah @lattner and @darthscsi suggested that. I was super happy to find this is already implemented in LLVM, and also in a pretty neat manner (much more efficient than the stuff I came up with :grin:)!